bmt/bmt.go

```diff
@@ -0,0 +1,571 @@
+	l := len(b)
+	if l <= 0 {
+		return 0, nil
+	}
+	s := self.segment
+	i := self.cur
+	count := (self.count + 1) / 2
+	need := self.count*self.size - self.cur*2*self.size
+	size := self.size
+	if need > size {
+		size *= 2
+	}
+	if l < need {
+		need = l
+	}
+	// calculate missing bit to complete current open segment
+	rest := size - len(s)
+	if need < rest {
+		rest = need
+	}
+	s = append(s, b[:rest]...)
+	need -= rest
+	//fmt.Printf("l: %v, s: %x, need: %v, size: %v, index: %v\n", l, s, need, size, self.cur)
+	// read full segments and the last possibly partial segment
+	for need > 0 && i < count-1 {
+		// push all finished chunks we read
+		self.writeSegment(i, s, self.depth)
+		need -= size
+		if need < 0 {
+			size += need
+		}
+		s = b[rest : rest+size]
+		rest += size
+		i++
+	}
+	//fmt.Printf("open segment %v len %v\n (data: %v)\n", i, len(s), hashstr(s))
+	self.segment = s
+	self.cur = i
+	// otherwise, we can assume len(s) == 0, so all buffer is read and chunk is not yet full
+	return l, nil
+}
+
+// reads from io.Reader and appends to the data to hash using Write
+// it reads so that chunk to hash is maximum length or reader reaches if EOF
+func (self *BMTHasher) ReadFrom(r io.Reader) (m int64, err error) {
+	self.getTree()
+	bufsize := self.size*self.count - self.size*self.cur - len(self.segment)
+	buf := make([]byte, bufsize)
+	var read int
+	for {
+		var n int
+		n, err = r.Read(buf)
+		read += n
+		if err == io.EOF || read == len(buf) {
+			hash := self.Sum(buf[:n])
+			if read == len(buf) {
+				err = NewEOC(hash)
+			}
+			break
+		}
+		if err != nil {
+			break
+		}
+		n, err = self.Write(buf[:n])
+		if err != nil {
+			break
+		}
+	}
+	return int64(read), err
+}
+
+func (self *BMTHasher) Reset() {
+	self.getTree()
+	self.blockLength = nil
+}
+
+// It implements the swarmHash interface
+func (self *BMTHasher) ResetWithLength(l []byte) {
+	self.Reset()
+	self.blockLength = l
+
+}
+
+// Release gives back the BMTree to the pool whereby it unlocks
+// it resets tree, segment and index
+func (self *BMTHasher) releaseTree() {
+	if self.bmt != nil {
+		n := self.bmt.leaves[self.cur]
+		for ; n != nil; n = n.parent {
+			n.unbalanced = false
+			if n.parent != nil {
+				n.root = false
+			}
+		}
+		self.pool.Release(self.bmt)
+		self.bmt = nil
+
+	}
+	self.cur = 0
+	self.segment = nil
+}
+
+type SegmentWriter interface {
+	Init(int) // initialises the segment writer with max int chunks
+	WriteSegment(int, []byte) error
+	Hash() []byte
+}
+
+func (self *BMTHasher) Hash() []byte {
+	return <-self.result
+}
+
+func (self *BMTHasher) Init(i int) {
+	self.getTree()
+	self.max = int32(i)
+}
+
+// implements the SegmentWriter interface ie allows for segments
+// to be written concurrently
+func (self *BMTHasher) WriteSegment(i int, s []byte) (err error) {
+	max := atomic.LoadInt32(&self.max)
+	if int(max) <= i {
+		return NewEOC(nil)
+	}
+	rightmost := i == int(max-1)
+	last := atomic.AddInt32(&self.max, 1) == max
```

Comment:
This look like a race: two goroutines atomically read `self.max`, both find `int(max) > i`, both potentially calculate `rightmost` to be true and race to set `self.segment`?